design-decisions/2018-11-14-Heating-Cooling-Representation.md

```diff
@@ -0,0 +1,179 @@
+Decision Template
+=================
+
+Background
+----------
+
+In order to evaluate expressions, usually one has to evaluate the
+expression terms first, then the expression itself is evaluated. As
+an example, `(1+2)+(3+4)` may be evaluated, in order, to the following
+equivalent patterns, where `[]` is a placeholder for an expression that was extracted from a larger one for evaluation, while `~>` separates expressions
+to be evaluated:
+```
+(1+2)+(3+4)
+1+2 ~> []+(3+4)  -- heating
+3 ~> []+(3+4)    -- rewriting
+3+(3+4)          -- cooling
+3+4 ~> 3+[]      -- heating
+7 ~> 3+[]        -- rewriting
+3+7              -- cooling
+10               -- rewriting
+```
+
+In many cases it is desirable to allow expressions to be evaluated in a fairly
+free order. Consider the following C example:
+```
+(f() + g()) + h()
+```
+where `f`, `g` and `h` have side effects. According to the standard, these
+functions can be called in any order, e.g. it is valid to call `g`, then `h`,
+then `f`.
+
+In all-path reachability proofs one attempts to show that, when starting from
+a given pattern, no matter how it is evaluated, one reaches a certain
+target pattern.
+
+In coinductive proofs, whenever there is a cycle in which a pattern can be
+rewritten, indirectly or not, to itself, anything can be proven.
+
+Problem/Questions
+-----------------
+
+How to encode the heating and cooling rules in kore, in such a way that there
+are no rewrite cycles, we can compute all patterns reachable from a given one,
+and we can prove that only those are reachable, while allowing any evaluation
+order for the heated/cooled terms when that's desirable.
+
+Decision: Use equations.
+------------------------
+
+Use rewriting rules like the following, where `~>` is a constructor,
+while `<k>` is only functional:
+
+```
+<k> a+b </k> = <k> a ~> []+b </k>    if a is not a value
+<k> a+b </k> = <k> b ~> a+[] </k>    if b is not a value
```

Comment:
If it's an equation, you need to drop the condition, as you might apply it from right to left (cool) when it's a value.